src/iosrv/sched.c

```diff
@@ -0,0 +1,539 @@
+{
+	struct sched_data	*data;
+	struct sched_cycle	*cycle;
+	int			 ret;
+
+	D_ALLOC_PTR(data);
+	if (data == NULL)
+		return ABT_ERR_MEM;
+
+	ret = ABT_sched_config_read(config, 2, &data->sd_event_freq,
+				    &data->sd_dx);
+	if (ret != ABT_SUCCESS) {
+		D_ERROR("Failed to read ABT sched config: %d\n", ret);
+		return ret;
+	}
+
+	cycle = &data->sd_cycle;
+	cycle->sc_age_net_bound[0] = SCHED_AGE_NET_MIN;
+	cycle->sc_age_net_bound[1] = SCHED_AGE_NET_MAX;
+	cycle->sc_age_nvme_bound[0] = SCHED_AGE_NVME_MIN;
+	cycle->sc_age_nvme_bound[1] = SCHED_AGE_NVME_MAX;
+
+	ret = ABT_sched_set_data(sched, (void *)data);
+	return ret;
+}
+
+static bool
+need_net_poll(struct sched_cycle *cycle)
+{
+	/* Need net poll to start new cycle */
+	if (!cycle->sc_cycle_started) {
+		D_ASSERT(cycle->sc_ults_tot == 0);
+		return true;
+	}
+
+	/* Need a nvme poll to end current cycle */
+	if (cycle->sc_ults_tot == 0)
+		return false;
+
+	/*
+	 * Need extra net poll when too many ULTs are processed in
+	 * current cycle.
+	 */
+	if (cycle->sc_age_net > cycle->sc_age_net_bound[1])
+		return true;
+
+	/* TODO: Take network request statistics into account */
+
+	return false;
+}
+
+static ABT_unit
+sched_pop_net_poll(struct sched_data *data, ABT_pool pool)
+{
+	struct dss_xstream	*dx = data->sd_dx;
+	struct sched_cycle	*cycle = &data->sd_cycle;
+	ABT_unit		 unit;
+	int			 ret;
+
+	if (!need_net_poll(cycle))
+		return ABT_UNIT_NULL;
+
+	cycle->sc_age_net = 0;
+	cycle->sc_age_nvme++;
+	if (cycle->sc_ults_tot == 0) {
+		D_ASSERT(!cycle->sc_cycle_started);
+		cycle->sc_new_cycle = 1;
+	}
+
+	/* XXX Enabled this check once dss_srv_handler() is revised */
+#if 0
+	/* The xstream without network poll ULT */
+	if (!dx->dx_comm)
+		return ABT_UNIT_NULL;
+#endif
+	ret = ABT_pool_pop(pool, &unit);
+	if (ret != ABT_SUCCESS) {
+		D_ERROR("XS(%d) failed to pop network poll ULT: %d\n",
+			dx->dx_xs_id, ret);
+		return ABT_UNIT_NULL;
+	}
+
+	return unit;
+}
+
+static bool
+need_nvme_poll(struct sched_cycle *cycle)
+{
+	/* Need net poll to start new cycle */
+	if (!cycle->sc_cycle_started) {
+		D_ASSERT(cycle->sc_ults_tot == 0);
+		return false;
+	}
+
+	/* Need nvme poll to end current cycle */
+	if (cycle->sc_ults_tot == 0)
+		return true;
+
+	/*
+	 * Need extra NVMe poll when too many ULTs are processed in
+	 * current cycle.
+	 */
+	if (cycle->sc_age_nvme > cycle->sc_age_nvme_bound[1])
+		return true;
+
+	/* TODO: Take NVMe I/O statistics into account */
+
+	return false;
+}
+
+static ABT_unit
+sched_pop_nvme_poll(struct sched_data *data, ABT_pool pool)
+{
+	struct dss_xstream	*dx = data->sd_dx;
+	struct sched_cycle	*cycle = &data->sd_cycle;
+	ABT_unit		 unit;
+	int			 ret;
+
+	if (!need_nvme_poll(cycle))
+		return ABT_UNIT_NULL;
+
+	D_ASSERT(cycle->sc_cycle_started);
+	cycle->sc_age_nvme = 0;
+	cycle->sc_age_net++;
+	if (cycle->sc_ults_tot == 0)
+		cycle->sc_cycle_started = 0;
+
+	/* Only main xstream (VOS xstream) has NVMe poll ULT */
+	if (!dx->dx_main_xs)
+		return ABT_UNIT_NULL;
```

Comment:
should this move above with need_nvme_poll() ? i.e. should sc_age_net/nvme being changed in this case?

Reply:
No, since we use same scheduler for both VOS xstream and helper xstream, the age needs be updated no matter if this is a real poll ULT is executed.